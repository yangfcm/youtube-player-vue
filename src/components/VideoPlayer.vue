<script setup lang="ts">
defineProps<{
  videoId: string
}>()
</script>

<template>
  <div class="app-player-container mb-2">
    <iframe
      :src="`https://www.youtube.com/embed/${videoId}`"
      alt="video"
      title="video player"
      class="app-player"
    >
    </iframe>
  </div>
</template>

<style scoped>
.app-player-container {
  position: relative;
  max-width: 100%;
  height: 0;
  overflow: hidden;
  background: #dcddde;
  padding-bottom: 75%;
  background: black;
}
@media screen and (min-width: 500px) {
  .app-player-container {
    padding-bottom: 55%;
  }
}
@media screen and (min-width: 1500px) {
  .app-player-container {
    padding-bottom: 45%;
  }
}
.app-player {
  width: 100%;
  height: 100%;
  position: absolute;
  border: none;
}
</style>
