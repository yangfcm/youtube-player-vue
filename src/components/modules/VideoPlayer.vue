<template>
  <div class="app-player-container">
    <iframe
      :src="videoSrc"
      alt="video"
      title="video player"
      class="app-player"
    ></iframe>
  </div>
</template>

<script>
export default {
  props: ["videoId"],
  computed: {
    videoSrc() {
      return `https://www.youtube.com/embed/${this.videoId}`;
    },
  },
};
</script>

<style scoped>
.app-player-container {
  position: relative;
  max-width: 100%;
  height: 0;
  overflow: hidden;
  background: #dcddde;
  padding-bottom: 62%;
}
.app-player {
  width: 100%;
  height: 100%;
  position: absolute;
}
</style>
