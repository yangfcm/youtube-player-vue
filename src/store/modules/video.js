import axios from "../../settings";
import { FETCH_VIDEOS, CATCH_VIDEO_ERROR, FETCH_VIDEO } from "../types";
import parseError from "../helpers/parseError";

const state = {
  videos: null,
  video: null,
  videoError: null,
};

const getters = {
  videoErrorMessage: (state) => {
    return parseError(state.videoError, "Failed to fetch video");
  },
};

const mutations = {
  FETCH_VIDEOS: (state, payload) => {
    state.videos = payload;
    state.videoError = null;
  },
  FETCH_VIDEO: (state, payload) => {
    state.video = payload;
    state.videoError = null;
  },
  CATCH_VIDEO_ERROR: (state, payload) => {
    state.videoError = payload;
  },
};

const actions = {
  /** Fetch a list of videos by filter */
  fetchVideos: async (context, filter, pageToken) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          ...filter,
          part: "snippet,statistics",
          maxResults: 15,
          pageToken,
        },
      });
      context.commit(FETCH_VIDEOS, response.data);
    } catch (err) {
      context.commit(CATCH_VIDEO_ERROR, err);
    }
  },

  /** Fetch a video by id */
  fetchVideo: async (context, videoId) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          part: "snippet,statistics",
          id: videoId,
        },
      });
      // console.log(response);
      context.commit(FETCH_VIDEO, response.data);
    } catch (err) {
      // console.log("error-==================", e);
      context.commit(CATCH_VIDEO_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
