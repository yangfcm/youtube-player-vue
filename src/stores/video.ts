import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { AsyncStatus } from '@/settings/types'
import { type VideoMetaSnippetStats, type VideoResponse } from './types'
import { fetchVideoAPI } from './api'
import { type AxiosResponse } from 'axios'

type VideoStore = {
  error: string
  status: AsyncStatus
  video: Record<string, VideoMetaSnippetStats>
}

export const useVideoStore = defineStore('video', () => {
  const videoStore = ref<VideoStore>({
    error: '',
    status: AsyncStatus.IDLE,
    video: {},
  })

  const status = computed(() => videoStore.value.status)
  const error = computed(() => videoStore.value.error)
  const video = computed(() => videoStore.value.video || {})

  const fetchVideo = async (videoId: string) => {
    try {
      videoStore.value.status = AsyncStatus.LOADING
      videoStore.value.error = ''
      const response: AxiosResponse<VideoResponse> = await fetchVideoAPI(videoId)
      const video = response.data.items[0]
      if (video) {
        videoStore.value.video = {
          ...(videoStore.value.video || {}),
          [video.id as string]: video,
        }
      }
      videoStore.value.status = AsyncStatus.SUCCESS
    } catch (err: any) {
      videoStore.value.status = AsyncStatus.FAIL
      videoStore.value.error = err.message
    }
  }

  return {
    videoState: videoStore,
    video,
    status,
    error,
    fetchVideo,
  }
})
