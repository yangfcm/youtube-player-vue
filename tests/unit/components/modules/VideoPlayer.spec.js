import { shallowMount } from "@vue/test-utils";
import VideoPlayer from "@/components/modules/VideoPlayer";
import { videoItem1 as video } from "../../store/fixtures/video";

describe("Test VideoPlayer component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(VideoPlayer, {
      propsData: {
        videoId: video.id,
      },
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("player should have correct src", () => {
    expect(wrapper.find(".app-player").attributes("src")).toBe(
      `https://www.youtube.com/embed/${video.id}`
    );
  });
});
