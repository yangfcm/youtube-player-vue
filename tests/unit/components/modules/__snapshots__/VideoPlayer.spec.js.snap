// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test VideoPlayer component should match snapshot 1`] = `<div class="app-player-container"><iframe src="https://www.youtube.com/embed/CacIlFaXph0" alt="video" title="video player" class="app-player"></iframe></div>`;
